File: swift/Okapi/Tests/OkapiTests/LdProofsTests.swift
//
// Created by Scott Phillips on 9/8/21.
//

import XCTest
import SwiftProtobuf
import Foundation

@testable import OkapiSwift

final class LdProofsTests: XCTestCase {
    func testGenerateCapabilityInvocationProofWithJCS() throws {
        let capabilityDictionary = ["@context": Google_Protobuf_Value(stringLiteral: "https://w3id.org/security/v2"),
                                    "target": Google_Protobuf_Value(stringLiteral: "urn:trinsic:wallets:noop"),
                                    "proof": Google_Protobuf_Value(structValue: Google_Protobuf_Struct(fields: [
                                        "created": Google_Protobuf_Value(stringLiteral:ISO8601DateFormatter().string(from: Date()))
                                    ]))
        ] as [String: Google_Protobuf_Value];
        var capabilityStruct: Google_Protobuf_Struct = Google_Protobuf_Struct();
        capabilityStruct.fields = capabilityDictionary;

        var request = Okapi_Keys_GenerateKeyRequest();
        request.keyType = Okapi_Keys_KeyType.ed25519;
        let response = try DidKey.generate(request: request);
        let signingKey = response.key.first { x in
            x.crv == "Ed25519"
        };

        var proofRequest = Okapi_Proofs_CreateProofRequest();
        proofRequest.document = capabilityStruct;
        proofRequest.key = signingKey!;
        proofRequest.suite = Okapi_Proofs_LdSuite.jcsEd25519Signature2020;

        _ = try LdProofs.createProof(request: proofRequest);
    }
}

File: dotnet/Tests/Okapi.Tests/LdProofsTests.cs
using System;
using System.Linq;
using FluentAssertions;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Newtonsoft.Json.Linq;
using Okapi.Keys;
using Okapi.Keys.V1;
using Okapi.Proofs;
using Okapi.Proofs.V1;
using Xunit;

namespace Okapi.Tests;

public class LdProofsTests
{
    [Fact]
    public void GenerateCapabilityInvocationProofWithJcs()
    {
        var capability = new JObject
        {
            { "@context", "https://w3id.org/security/v2" },
            { "target", "urn:trinsic:wallets:noop" },
            {
                "proof", new JObject
                {
                    { "created", DateTime.UtcNow.ToString("s") }
                }
            }
        };

        var key = DIDKey.Generate(new GenerateKeyRequest { KeyType = KeyType.Ed25519 });
        var signingKey = key.Key.First(x => x.Crv == "Ed25519");

        var signedCapability = LDProofs.CreateProof(new CreateProofRequest
        {
            Key = signingKey,
            Document = capability.ToStruct(),
            Suite = LdSuite.Jcsed25519Signature2020
        });

        signedCapability.Should().NotBeNull();
        signedCapability.SignedDocument.Should().NotBeNull();
    }
}

public static class JTokenExtensions
{
    /// <summary>
    ///     Converts to struct.
    /// </summary>
    /// <param name="jobj">The jobj.</param>
    /// <returns></returns>
    public static Struct ToStruct(this JToken token)
    {
        return JsonParser.Default.Parse<Struct>(token.ToString());
    }
}